applications/MappingApplication/custom_searching/interface_communicator.cpp

```diff
@@ -261,51 +262,66 @@ void InterfaceCommunicator::ConductLocalSearch(const Communicator& rComm)
 
-    KRATOS_ERROR_IF(mSearchRadius < 0.0) << "Search-Radius has to be larger than 0.0!"
-        << std::endl;
+    KRATOS_ERROR_IF(mSearchRadius < 0.0) << "Search-Radius has to be larger than 0.0!" << std::endl;
 
     int sum_num_results = 0;
     int sum_num_searched_objects = 0;
 
     if (num_interface_obj_bin > 0) { // this partition has a bin structure
-        InterfaceObjectConfigure::ResultContainerType neighbor_results(num_interface_obj_bin);
-        auto interface_obj(Kratos::make_shared<InterfaceObject>(array_1d<double, 3>(0.0)));
 
-        for (auto& r_interface_infos_rank : mMapperInterfaceInfosContainer) { // loop the ranks
-            // #pragma omp parallel for // TODO this requires to make some things thread-local!
-            // it makes more sense to omp this loop even though it is not the outermost one ...
-            for (IndexType i=0; i<r_interface_infos_rank.size(); ++i) {
-                sum_num_searched_objects++;
+        struct SearchTLS {
+            SearchTLS(std::size_t MaxNeighborResults) : mMaxNeighborResults(MaxNeighborResults) {}
+
+            // the IndexPartition uses the CopyConstructor to create the thread local storage
+            // hence using it to initialize the members
+            SearchTLS(const SearchTLS& rOther)
+            {
+                mNeighborResults.resize(rOther.mMaxNeighborResults);
+                mInterfaceObject = Kratos::make_shared<InterfaceObject>(array_1d<double, 3>(0.0));
+            }
 
-                auto& r_interface_info = r_interface_infos_rank[i];
+            InterfaceObjectConfigure::ResultContainerType mNeighborResults;
+            Kratos::shared_ptr<InterfaceObject> mInterfaceObject;
 
-                interface_obj->Coordinates() = r_interface_info->Coordinates();
-                double search_radius = mSearchRadius; // reset search radius // TODO check this
+        private:
+            std::size_t mMaxNeighborResults;
+        };
+
+        for (auto& r_interface_infos_rank : mMapperInterfaceInfosContainer) { // loop the ranks
+            sum_num_searched_objects += r_interface_infos_rank.size();
+            // intentionally not the outermost loop is parallelized as this one can be large
+            sum_num_results += IndexPartition<std::size_t>(r_interface_infos_rank.size()).for_each<SumReduction<int>>(
+                SearchTLS(num_interface_obj_bin), [
+                &r_interface_infos_rank,
+                num_interface_obj_bin,
+                this](const std::size_t Index, SearchTLS& rTLS) {
+                auto& r_interface_info = r_interface_infos_rank[Index];
+
+                rTLS.mInterfaceObject->Coordinates() = r_interface_info->Coordinates();
 
                 // reset the containers
-                auto results_itr = neighbor_results.begin();
+                auto results_itr = rTLS.mNeighborResults.begin();
 
                 const SizeType number_of_results = mpLocalBinStructure->SearchObjectsInRadius(
```

Comment:
@roigcarlo I keep using this function as this way I can make it parallel quite easy
I checked the implementation, I think it is threadsafe, can you please confirm?

I also checked the [DEM search](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/DEMApplication/custom_utilities/omp_dem_search.h#L231-L251), there they also use this function in an omp loop
(@maceligueta can you confirm that the `OMP_DEMSearch` is actually used? I think so but not 100% sure. Thx!)

Reply:
Yes, OMP_DEMSearch is used as the stardard search engine.